apps/emqx_durable_storage/src/emqx_ds_new_streams.erl

```diff
@@ -0,0 +1,405 @@
+%% `emqx_durable_storage.new_streams_batch_size' application
+%% environment variable), with a cooldown in between (configurable by
+%% `emqx_durable_storage.new_streams_cooldown').
+%%
+%% This is done to avoid a storm of `emqx_ds:get_streams' calls from
+%% the clients.
+-module(emqx_ds_new_streams).
+
+-behaviour(gen_statem).
+
+%% API:
+-export([start_link/1, where/1]).
+-export([watch/2, unwatch/2]).
+-export([notify_new_stream/2, set_dirty/1]).
+
+-export_type([watch/0]).
+
+%% Local API and RPC targets
+-export([local_notify_new_stream/2, local_set_dirty/1]).
+
+%% behavior callbacks:
+-export([callback_mode/0, init/1, handle_event/4]).
+
+%% For testing:
+-export([list_subscriptions/1]).
+
+-include_lib("emqx_durable_storage/include/emqx_ds.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-opaque watch() :: reference().
+
+-define(via(DB), {via, gproc, {n, l, {?MODULE, DB}}}).
+
+%% States:
+-define(clean, clean).
+-define(dirty, dirty).
+
+-type state() :: ?clean | ?dirty.
+
+%% Timeouts:
+-define(dirty_loop, dirty_loop).
+
+-record(d, {
+    db :: emqx_ds:db(),
+    subs :: ets:tid(),
+    trie :: ets:tid(),
+    dispatch_iterator
+}).
+
+-type data() :: #d{}.
+
+-record(sub, {id, tf, pid, active = false}).
+
+-type sub() :: #sub{
+    id :: watch(),
+    tf :: emqx_ds:topic_filter(),
+    pid :: pid(),
+    active :: boolean()
+}.
+
+%% Calls and casts:
+-record(watch_req, {topic_filter :: emqx_ds:topic_filter()}).
+-record(unwatch_req, {ref :: watch()}).
+-record(notify_req, {topic_filter :: emqx_ds:topic_filter()}).
+-record(list_subs_req, {}).
+
+%%================================================================================
+%% API functions
+%%================================================================================
+
+-spec where(emqx_ds:db()) -> pid() | undefined.
+where(DB) ->
+    gproc:where({n, l, {?MODULE, DB}}).
+
+%% @doc Process that calls this function will receive messages of type
+%% `#new_stream_event{subref = Ref}' when new streams matching the
+%% topic filters are created in the durable storage.
+%%
+%% Note: this function is not idempotent.
+-spec watch(emqx_ds:db(), emqx_ds:topic_filter()) -> {ok, watch()} | {error, badarg}.
+watch(DB, TopicFilter) ->
+    gen_server:call(?via(DB), #watch_req{topic_filter = TopicFilter}).
+
+-spec unwatch(emqx_ds:db(), watch()) -> ok.
+unwatch(DB, Ref) ->
+    gen_statem:call(?via(DB), #unwatch_req{ref = Ref}).
+
+%% @doc Broadcast notification about appearance of new stream(s) to
+%% all nodes.
+-spec notify_new_stream(emqx_ds:db(), emqx_ds:topic_filter()) -> ok.
+notify_new_stream(DB, TF) ->
+    emqx_ds_new_streams_proto_v1:notify([node() | nodes()], DB, TF).
```

Comment:
Don't we need running_nodes here?

Reply:
In theory, yes. But I don't like the kludge of using `mria:running_nodes()`. It's mostly used as a workaround to avoid sending rpc **calls** to nodes that can't handle them (e.g. rshell), it's less relevant for casts.